src/Compilers/CSharp/Portable/Symbols/NonMissingAssemblySymbol.cs

```diff
@@ -85,7 +85,7 @@ public sealed override NamespaceSymbol GlobalNamespace
         /// <param name="digThroughForwardedTypes">
         /// Take forwarded types into account.
         /// </param>
-        internal sealed override NamedTypeSymbol LookupTopLevelMetadataTypeWithCycleDetection(ref MetadataTypeName emittedName, ConsList<AssemblySymbol> visitedAssemblies, bool digThroughForwardedTypes)
+        internal sealed override NamedTypeSymbol LookupTopLevelMetadataTypeWithCycleDetection(ref MetadataTypeName emittedName, ConsList<AssemblySymbol> visitedAssemblies, bool digThroughForwardedTypes, bool returnNullForMissing = false)
```

Comment:
>bool returnNullForMissing = false [](http://example.com/codeflow?start=203&length=33)

I think we should investigate if we could do better than a conditional optimization and delegate creation of missing symbols to specific consumers that actually need them. Likely there aren't that many of them. Also, I would want us to limit scenarios that affect content of the ```_emittedNameToTypeMap``` that this function is using. This cache was not meant to cache results of random lookup requests. It was meant to only serve internal compiler requests to resolve type references during metadata import and in some other similar scenarios. There was no intent to cache results of lookup requests made by external consumers, because that can lead to a significant cache bloat and other negative effects.